batch/batch/aioclient.py

```diff
@@ -152,36 +287,31 @@ def __init__(self, client, id, attributes):
         if pvc_size:
             doc['pvc_size'] = pvc_size
 
-        j = await self._client._post('/jobs/create', json=doc)
+        self._job_docs.append(doc)
 
-        return Job(self,
-                   j['job_id'],
-                   attributes=j.get('attributes'),
-                   parent_ids=j.get('parent_ids', []))
+        j = Job.unsubmitted_job(self, self._job_idx, attributes, parent_ids)
+        self._jobs.append(j)
```

Comment:
I don't quite understand why _jobs should be populated at this stage; you already have _job_docs, which is the objet representation of the eventual entry in the db. Then you can at a glance tell how many jobs you've submitted (`len(j._jobs)`), and avoid a linear search to update each of those jobs below after submission.

Reply:
To be clear: I obviously can't see your full plan, and so this model may fully make sense.

Reply:
Namely, Job.unsubmitted_job holds precisely the same data as the doc itself. No need to introduce additional abstractions imo

Reply:
There is a major design constraint on the batch client. We want the user to be able to say `batch.create_job()` and receive a job object that they can use to query the job and use it to build the DAG. We have to have some way to change the state of the Job from the unsubmitted state to the submitted state. There's no way to do that without a linear search to update each job. @danking and I thought about this yesterday and this is what we concluded. I'd be happy to hear a concrete proposal on how to keep the current interface without doing a O(N) pass over the data.

Reply:
The only way to do this not in O(N) time was how I had it before where everything was linked to the batch id and I set the batch id from None to an actual id. you didn't like this interface and I agreed. @danking and I thought doing a loop over the array would pale in comparison to the time to do the network request and insert the new records into the database.

Reply:
Agreed that loop isn't necessary. I'm not clear that you're actually submitting the 'jobs' aka job_docs to /create/batch, but I may be missing it.

So in my mental model:
1) You build a job_doc graph : excellent
2) You submit this graph and instantiate batch with it. This graph is therefore considered "submitted".
3) If you want to find out if a node in the graph is complete, you call a function on Batch, which does a query on a specific id.

  - The ID is the index in the jobs array

Alternatively, if you want to keep the graph in BatchBuilder, just take a pointer to it as self._jobs, one step.

If there is major benefit to having 1 additional object (with methods) per node, then it should be used, but in general, constructing that many objects seems like it will cost memory, garbage collection time, without much benefit.

I understand this may not be the Hail way. I like minimal OO.

Reply:
As we talked, this is not needed for this PR, we'll talk about it as a team :)